File: app/src/main/java/adong/org/modiproject/service/RetrofitService.java
package adong.org.modiproject.service;

import java.util.List;

import adong.org.modiproject.data.Token;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.logging.HttpLoggingInterceptor;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

public class RetrofitService {
    private final static String BASE_URL = "http://112.158.99.134:3000";

    private static Retrofit retrofit = null;

    public static Retrofit getClient() {
        HttpLoggingInterceptor interceptor = new HttpLoggingInterceptor();
        interceptor.setLevel(HttpLoggingInterceptor.Level.BODY);
        OkHttpClient client = new OkHttpClient.Builder()
                .addInterceptor(interceptor)
                .build();

        retrofit = new Retrofit.Builder()
                .baseUrl(BASE_URL)
                .addConverterFactory(GsonConverterFactory.create())
                .client(client)
                .build();

        return retrofit;
    }

    public static Retrofit getLoginRetrofit() {
        OkHttpClient.Builder httpClient = new OkHttpClient.Builder();
        httpClient.addInterceptor(chain -> {
            Request original = chain.request();
            Token token = Token.last(Token.class);
            Request request = original.newBuilder()
                    .header("authorization", token.getData())
                    .method(original.method(), original.body())
                    .build();
            return chain.proceed(request);
        });

        OkHttpClient client = httpClient.build();
        retrofit = new Retrofit.Builder()
                .baseUrl(BASE_URL)
                .addConverterFactory(GsonConverterFactory.create())
                .client(client)
                .build();

        return retrofit;
    }

}

File: app/src/main/java/adong/org/modiproject/service/RetrofitService.kt
package adong.org.modiproject.service

import android.text.TextUtils
import okhttp3.Interceptor
import okhttp3.OkHttpClient
import okhttp3.Response
import okhttp3.logging.HttpLoggingInterceptor
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory

class RetrofitService {
    val BASE_URL = "http://112.158.99.134:3000"

    var retrofit : Retrofit? = null
    val httpclient = OkHttpClient.Builder()
    val builder = Retrofit.Builder()
            .baseUrl(BASE_URL)
            .addConverterFactory(GsonConverterFactory.create())

    fun <T> creatService(serviceClass : Class<T>) : T {
        val interceptor = HttpLoggingInterceptor()
        interceptor.level = HttpLoggingInterceptor.Level.BODY
        httpclient.addInterceptor(interceptor)
        retrofit = builder
                .client(httpclient.build())
                .build()
        return retrofit!!.create(serviceClass)
    }

    fun <T> creatService(serviceClass : Class<T>, authToken : String) : T{
        val interceptorlevel = HttpLoggingInterceptor()
        interceptorlevel.level = HttpLoggingInterceptor.Level.BODY
        if (!TextUtils.isEmpty(authToken)){
            val interceptor = AuthenticationInterceptor(authToken)
            if (!httpclient.interceptors().contains(interceptor)){
                httpclient.addInterceptor(interceptor)
                httpclient.addInterceptor(interceptorlevel)
                builder.client(httpclient.build())
                retrofit = builder.build()
                return retrofit!!.create(serviceClass)
            }
        }
        return retrofit!!.create(serviceClass)
    }
}

class AuthenticationInterceptor(val authToken : String) : Interceptor{

    override fun intercept(chain: Interceptor.Chain?): Response {
        val original = chain!!.request()

        val builder = original.newBuilder()
                .header("Authorization", authToken)

        val request = builder.build()

        return chain.proceed(request)
    }

}
